### PR TITLE
Feature login signup

### PR DESCRIPTION
**Pull request by zachwreed**
**This PR includes:**
- Signup UI
- Signup API endpoint with password hash & salt
- Login UI
- Signup API endpoint with password hash & salt check and token return
- Fixes routing and redirection based on user state
- Middleware to validate username and token in LocalStorage
- LocalStorage API check to validate user on subsequent sessions
- Removes Base.jsx due to routing issues and points index.js entry to App.jsx

### CODE REVIEW

**Comment by mobilemelody** [comment]
@zachwreed I think it'd be good if you could resolve the conflicts with `development` first before we review?

**Review by mobilemelody — COMMENTED** [review]
Looks good to me. Just a few comments/questions:
- Once logged in, the login form flashes briefly on the homepage before displaying all the games
- Would it make sense to add the game ID to the URL when on an individual game page? Also, when you reload a game page, it goes back to the all games page
- It looks like some of the linting errors that were fixed in PR #23 were reverted back? For example, the indentation in `client/src/reducers/reviews.js`

**Comment by zachwreed** [comment]
> * It looks like some of the linting errors that were fixed in PR #23 were reverted back? For example, the indentation in `client/src/reducers/reviews.js`

What linter are you using? Do you have indents set to 2 or 4 spaces?

**Comment by zachwreed** [comment]
> * Once logged in, the login form flashes briefly on the homepage before displaying all the games

Yes the login flash is an issue I believe is related so some of the state change issues you and Jed have mentioned. I would like to deal with those issues all in one branch.

> * Would it make sense to add the game ID to the URL when on an individual game page? Also, when you reload a game page, it goes back to the all games page

I agree, the game page should have the id added to the URL as well as a few other changes, I think that is better suited for a different branch to address some refactoring to the Game.jsx page.

**Comment by zachwreed** [comment]
> * It looks like some of the linting errors that were fixed in PR #23 were reverted back? For example, the indentation in `client/src/reducers/reviews.js`

Alright I think I took care of that. My linter doesn't show an error - let me know if that looks better for you.

File: api/middleware/index.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            message: "Header is required" 
+        })
+    }
+
+    // Split token from bearer header
+    const bearer = bearer_header.split(' ')
+    const bearer_token = bearer[1] // <token>
+
+    // Verify token
+    jwt.verify(bearer_token, 'secret', function(err, decoded) {
```
Comment (Jedeu) [inline api/middleware/index.js]: Can we camelCase the variable names please? Just going along with the Javascript style

File: api/middleware/index.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    if (decoded.username !== from) {
+    return res.status(400).set({
+        "Content-Type": "application/json",
+      }).send({
+        status: 'error',
+        message: 'invalid username-token combination',
+      })
+    }
+
+    console.log("here")
```
Comment (Jedeu) [inline api/middleware/index.js]: Can we remove this console.log please?

File: api/middleware/index.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
+module.exports = function(req, res, next) {
+    var jwt = require('jsonwebtoken')
```
Comment (Jedeu) [inline api/middleware/index.js]: Let's make this a const if you don't plan on updating the value

File: api/routes/users.js [hunk 4]
```diff
@@ -3,6 +3,134 @@ const router = express.Router();
 const db = require('../db');
 const dbUtils = require('../utils/db.js');
 const apiUtils = require('../utils/api.js');
+var jwt = require('jsonwebtoken');
```
Comment (Jedeu) [inline api/routes/users.js]: Let's also make this a const if you don't plan on changing it

File: api/routes/users.js [hunk 5]
```diff
@@ -3,6 +3,134 @@ const router = express.Router();
...
+        }).send({
+          status: 'error',
+          message: err
+        });
+      }
+
+      // Create token with username
+      var token = jwt.sign({username: result.rows[0].username}, secret)
+
+      console.log("success")
```
Comment (Jedeu) [inline api/routes/users.js]: Could we remove this console.log here if we're done debugging?

File: client/src/components/AddGame.jsx [hunk 6]
```diff
@@ -139,24 +139,24 @@ class AddGameForm extends React.Component {
...
-          },
-        })
+        .then(res => res.json())
         .then(res => {
-          // Save values to database
-          values.image = url;
-          this.saveToDB(values);
+          let returnData = res.data.returnData;
+          let signedRequest = returnData.signedRequest;
+          let url = returnData.url;
```
Comment (Jedeu) [inline client/src/components/AddGame.jsx]: You can shorten this to be:

```
const { signedRequest, url } = res.data.returnData
```

File: client/src/components/App.jsx [hunk 7]
```diff
@@ -17,62 +13,75 @@ import "mdbreact/dist/css/mdb.css";
...
     super(props)
     this._userLogout = this._userLogout.bind(this)
   }
 
-  componentDidMount(){
-    // Add method to check if user is already logged in by checking token in localStorage
+  componentDidMount() {
+    var token = localStorage.getItem('token')
+    var username = localStorage.getItem('username')
+
```
Comment (Jedeu) [inline client/src/components/App.jsx]: let's use `const` here instead of `var`

File: client/src/components/App.jsx [hunk 8]
```diff
@@ -17,62 +13,75 @@ import "mdbreact/dist/css/mdb.css";
...
-  _userLogout(){
+  _userLogout() {
     this.props.userLogout()
   }
 
   render() {
 
     const { user } = this.props
-    const { game } = this.props
+    // console.log(user)
```
Comment (Jedeu) [inline client/src/components/App.jsx]: Let's remove this console.log if you don't need it anymore

File: client/src/components/Game.jsx [hunk 9]
```diff
@@ -2,79 +2,108 @@ import React from 'react';
...
+  constructor(props) {
+    super(props)
+  }
+
   render() {
 
-    const { game } = this.props
-    // const { review }
+    const { game, user } = this.props
+    console.log(game)
```
Comment (Jedeu) [inline client/src/components/Game.jsx]: Let's remove the console.log if we don't need it

File: client/src/components/Game.jsx [hunk 10]
```diff
@@ -2,79 +2,108 @@ import React from 'react';
...
+    var body
```
Comment (Jedeu) [inline client/src/components/Game.jsx]: Let's use let instead of `var`

**Review by Jedeu — APPROVED** [review]
